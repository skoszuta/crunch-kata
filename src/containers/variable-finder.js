import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import getNodeByPath from 'utils/get-node-by-path';
import getVariableName from 'utils/get-variable-name';
import Chip, { ChipsList } from 'components/chip';

import backspaceIcon from 'images/backspace-icon.svg';

const LastNode = styled.li`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  flex-wrap: nowrap;

  margin-top: 0.5rem;
`;

const BackspaceButton = styled.button`
  margin-left: 0.5rem;
  outline: none;
  width: 24px;
  height: 24px;
  border: 0;
  background-color: transparent;
  background-image: url(${backspaceIcon});
  cursor: pointer;
`;

class VariableFinder extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      path: []
    };
  }

  onNodeSelectChange({ target: { value } }) {
    this.setState({
      path: [...this.state.path, value]
    });
  }

  onBackspaceClick() {
    this.setState({
      path: this.state.path.slice(0, -1)
    });
  }

  renderLastNode() {
    const { index, order } = this.props.variables;
    const lastNode = getNodeByPath(this.state.path, order);
    const backspace = this.state.path.length ? <BackspaceButton onClick={this.onBackspaceClick.bind(this)}/> : '';

    if (typeof lastNode === 'string') {
      return [
        <Chip key="chip">{getVariableName(index[lastNode])}</Chip>,
        <LastNode key="backspace">{backspace}</LastNode>
      ];
    } else {
      const options = (typeof lastNode === 'object') ? Object.keys(lastNode) : lastNode.keys();
      const optionsDOM = options.map((option) => <option key={option}>{option}</option>);

      return (
        <LastNode>
          <select onChange={this.onNodeSelectChange.bind(this)}>
            <option value={void 0}>Select a key...</option>
            {optionsDOM}
          </select>
          {backspace}
        </LastNode>
      )
    }
  }

  renderPath() {
    return ['Root', ...this.state.path].map((segment, index) => (
      <Chip arrow={true} key={index}>{segment}</Chip>
    ));
  }

  render() {
    const { index, order } = this.props.variables;

    return (
      <div>
        <ChipsList>
          {this.renderPath()}
          {this.renderLastNode()}
        </ChipsList>
      </div>
    )
  }

  getVariableByPath(path) {
    const { index, order } = this.props.variables;
    return index[getNodeByPath(path, order)];
  }
}

function mapStateToProps({ variables }) {
  return { variables };
}

export default connect(mapStateToProps, null)(VariableFinder);
