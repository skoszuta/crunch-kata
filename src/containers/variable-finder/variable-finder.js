import React from 'react';
import { connect } from 'react-redux';

import getNodeByPath from '../../utils/get-node-by-path';
import getVariableName from '../../utils/get-variable-name';

class VariableFinder extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      path: []
    };
  }

  onNodeSelectChange({ target: { value } }) {
    this.setState({
      path: [...this.state.path, value]
    });
  }

  onBackButtonClick() {
    this.setState({
      path: this.state.path.slice(0, -1)
    });
  }

  renderLastNode() {
    const { index, order } = this.props.variables;
    const lastNode = getNodeByPath(this.state.path, order);


    if (typeof lastNode === 'string') {
      return getVariableName(index[lastNode]);
    } else {
      const options = (typeof lastNode === 'object') ? Object.keys(lastNode) : lastNode.keys();
      const optionsDOM = options.map((option) => <option key={option}>{option}</option>);

      return (
        <span>
          <select onChange={this.onNodeSelectChange.bind(this)}>
            <option value={void 0}>Select an index...</option>
            {optionsDOM}
          </select>
        </span>
      )
    }
  }

  render() {
    const { index, order } = this.props.variables;
    const backButton = this.state.path.length ? <button onClick={this.onBackButtonClick.bind(this)}>Back</button> : '';

    return (
      <div>
        <div>
          Root -> {this.state.path.join(' -> ')} {this.state.path.length ? ' -> ' : ''}{this.renderLastNode()}{backButton}
        </div>
      </div>
    )
  }

  getVariableByPath(path) {
    const { index, order } = this.props.variables;
    return index[getNodeByPath(path, order)];
  }
}

function mapStateToProps({ variables }) {
  return { variables };
}

export default connect(mapStateToProps, null)(VariableFinder);
