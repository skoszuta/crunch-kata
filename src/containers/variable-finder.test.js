import fs from 'fs';
import React from 'react';
import { mount, render } from 'enzyme';
import 'jest-styled-components';

import { VariableFinder } from './variable-finder';

const { index } = JSON.parse(fs.readFileSync('json/variables.json', 'utf8'));
const { graph: order } = JSON.parse(fs.readFileSync('json/order.json', 'utf8'));

describe('VariableFinder', () => {
  const variables = { index, order };

  describe('snapshot', () => {
    it('should render correctly', () => {
      const wrapper = render(
        <VariableFinder variables={variables} />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    const validPathState = { path: [2, 'Demographics and Technographics', 3] };
    let wrapper;

    beforeEach(() => {
      wrapper = mount(
        <VariableFinder variables={variables} />
      );
    });

    it('should add a segment on select change', () => {
      const select = wrapper.find('select').at(0);
      select.prop('onChange')({ target: { value: 0 } });
      wrapper.update();

      expect(wrapper.find('li').length).toBe(3);
    });

    it('should hide the select when a variable is reached', () => {
      wrapper.setState(validPathState);
      wrapper.update();

      expect(wrapper.find('select').length).toBe(0);
    });

    it('should show a delete button when the path is not empty', () => {
      expect(wrapper.find('button').length).toBe(0);

      wrapper.setState(validPathState);
      wrapper.update();

      expect(wrapper.find('button').length).toBe(1);
    });
  });
});
