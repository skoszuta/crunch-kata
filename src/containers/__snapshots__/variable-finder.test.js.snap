// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariableFinder snapshot should render correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 0.5rem;
}

.c2 {
  display: inline-block;
  border-radius: 50px;
  padding: 0.5rem 1rem;
  font-size: 0.85rem;
  color: #f6f6f6;
  background-color: #666;
}

.c3 {
  display: inline-block;
  width: 24px;
  height: 24px;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  background-image: url(test-file-stub);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  margin-top: 0.5rem;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        Root
      </div>
      <div
        class="c3"
      />
    </li>
    <li
      class="c4"
    >
      <select>
        <option>
          Select a key...
        </option>
        <option>
          0
        </option>
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
        <option>
          4
        </option>
        <option>
          5
        </option>
      </select>
    </li>
  </ul>
</div>
`;
